### PR TITLE
Improve CryptoConfig.CreateFromName performance (from 10% to x20)

### PR DESCRIPTION
**Pull request by adamsitnik**
I was working on making it easier for CoreFX team members to use ASP.NET Perf Lab to run performance tests under high load.

Then I saw #39600 which was a perfect candidate for sample for the docs: we had some nasty locks that were making the perf bad. So I created sample docs based on this example: https://github.com/aspnet/Benchmarks/pull/1284 

After taking a brief look at the existing code, I could not resist rewriting it and make it even faster than @VladimirKhvostov did in #39600 

ASP.NET Perf lab test result (described in https://github.com/aspnet/Benchmarks/pull/1284):

Before:

```log
RequestsPerSecond:           43,330
Max CPU (%):                 95
WorkingSet (MB):             166
Avg. Latency (ms):           5.91
```

flamegraph_filtered

After:

```log
RequestsPerSecond:           120,801
Max CPU (%):                 90
WorkingSet (MB):             166
Avg. Latency (ms):           2.27
```

flamegraph_filtered_after

Some [microbenchmark](https://github.com/dotnet/performance/blob/master/src/benchmarks/micro/corefx/System.Security.Cryptography/Perf.CryptoConfig.cs) results:

|         Method |     Toolchain |      name |       Mean | Ratio | Allocated Memory/Op |
|--------------- |-------------- |---------- |-----------:|------:|--------------------:|
| CreateFromName |         net48 |       RSA |   765.4 ns |  1.00 |               802 B |
| CreateFromName | netcoreapp3.0 |       RSA | 3,681.8 ns |  4.81 |               688 B |
| CreateFromName |       this PR |       RSA |   278.1 ns |  0.36 |               192 B |
|                |               |           |            |       |                     |
| CreateFromName |         net48 |    SHA512 |   605.7 ns |  1.00 |              1316 B |
| CreateFromName | netcoreapp3.0 |    SHA512 |   770.7 ns |  1.27 |               424 B |
| CreateFromName |       this PR |    SHA512 |   545.3 ns |  0.90 |               128 B |
|                |               |           |            |       |                     |
| CreateFromName |         net48 | X509Chain | 4,188.9 ns |  1.00 |               682 B |
| CreateFromName | netcoreapp3.0 | X509Chain | 3,269.7 ns |  0.78 |               464 B |
| CreateFromName |       this PR | X509Chain |   228.1 ns |  0.05 |                88 B |

@danmosemsft this was one of the cases where Full .NET Framework was faster than Core. I made sure that Core is faster now.

The key changes:

* remove the locks, use ConcurrentDictionary instead
* `Type.GetType(string name)` is very expensive. Every type is now loaded once and then saved to the dictionary. This was the biggest improvement
* one dictionary for name to name mapping, one for name to type mapping (previously there was one for string to object)
* Fast path for creating config without providing any extra arguments

### CODE REVIEW

**Comment by stephentoub** [comment]
@adamsitnik, how do your changes relate to https://github.com/dotnet/corefx/pull/39600?  I've not reviewed yours, but from the description it seems duplicative.

**Comment by GrabYourPitchforks** [comment]
Are any real-world E2E scenarios improved by this change? I'd be surprised to learn that this function is the bottleneck in any legitimate use of cryptography.

**Comment by VladimirKhvostov** [comment]
@GrabYourPitchforks this method is used in the JWT token validation code path, which caused several LSIs in other DevOps. See #39600 for details.

**Comment by GrabYourPitchforks** [comment]
@VladimirKhvostov I saw that, but it seems the conclusion from that thread was to change the JWT verification implementation to call a different method. To an extent this matches recommended practice: the CryptoConfig class is really intended for use only when you need agility. JWT verification shouldn't be bouncing through this class because the JWT validation library should be hardcoding the list of supported algorithms; it doesn't require the type of agility the CryptoConfig class is intended to provide.

**Comment by danmosemsft** [comment]
cc @krwq

**Comment by krwq** [comment]
@adamsitnik thanks! I think next steps would be to not call CryptoConfig.CreateFromName at all and just use algorithm directly. That would avoid any lookup at all

File: src/System.Security.Cryptography.Algorithms/src/System/Security/Cryptography/CryptoConfig.cs [hunk 1]
```diff
@@ -2,520 +2,265 @@
...
                 throw new ArgumentException(SR.Cryptography_AlgorithmTypesMustBeVisible, nameof(algorithm));
+            }
             if (names == null)
+            {
                 throw new ArgumentNullException(nameof(names));
+            }
 
-            string[] algorithmNames = new string[names.Length];
-            Array.Copy(names, 0, algorithmNames, 0, algorithmNames.Length);
+            string[] safeCopy = names.Length == 1 ? new string[1] { names[0] } : names.ToArray();
```
Comment (krwq) [inline src/System.Security.Cryptography.Algorithms/src/System/Security/Cryptography/CryptoConfig.cs]: Does `new string[1] { names[0] }` actually make that difference? If so, could we perhaps improve `ToArray` instead?

File: src/System.Security.Cryptography.Algorithms/src/System/Security/Cryptography/CryptoConfig.cs [hunk 2]
```diff
@@ -640,5 +389,265 @@ private static void EncodeSingleOidNum(uint value, byte[] destination, ref int i
...
+
+            typeNameToType["AesManaged"] = AesManagedType;
+            typeNameToType["System.Security.Cryptography.AesManaged"] = AesManagedType;
+
+            // Add the other hash algorithms introduced with XML Encryption
+            typeNameToType["http://www.w3.org/2001/04/xmlenc#sha256"] = SHA256DefaultType;
+            typeNameToType["http://www.w3.org/2001/04/xmlenc#sha512"] = SHA512DefaultType;
+
+
+            typeNameToType["http://www.w3.org/2001/04/xmlenc#aes128-cbc"] = RijndaelManagedType;
```
Comment (krwq) [inline src/System.Security.Cryptography.Algorithms/src/System/Security/Cryptography/CryptoConfig.cs]: @bartonjs should we be using `AesManagedType` directly here?

File: src/System.Security.Cryptography.Algorithms/src/System/Security/Cryptography/CryptoConfig.cs (hunk 1 above)
Comment (VladimirKhvostov) [reply src/System.Security.Cryptography.Algorithms/src/System/Security/Cryptography/CryptoConfig.cs]: I suspect CryptoConfig.AddAlgorithm most often called with 1 argument. We can either special case this - not create safeCopy if names.Length == 1 or even create an overload AddAlgorithm, which would take string name instead string names[].

**Comment by krwq** [comment]
I have some security concerns about this fix (and also original PR) - will need to discuss it with @bartonjs since he's more familiar with how this is being usage. I'll take this offline

File: src/System.Security.Cryptography.Algorithms/src/System/Security/Cryptography/CryptoConfig.cs (hunk 1 above)
Comment (stephentoub) [reply src/System.Security.Cryptography.Algorithms/src/System/Security/Cryptography/CryptoConfig.cs]: This is definitely not worth extra code/complexity.  Just do:
```C#
string[] algorithmNAmes = names.AsSpan().ToArray();
```
It's simpler and its perf will be just fine; I cannot believe any perf difference will be measurable in real-world usage, especially considered we're about to do cryptography!

(If you really believed Length == 1 was special and warranted optimization, you'd want to special case everything for the rest of the method, too; there's no point allocating another array, looping with foreach twice, etc., when you can just operate on the sole value.)

File: src/System.Security.Cryptography.Algorithms/src/System/Security/Cryptography/CryptoConfig.cs [hunk 3]
```diff
@@ -640,5 +389,265 @@ private static void EncodeSingleOidNum(uint value, byte[] destination, ref int i
...
+            // string SHA384CryptoSerivceProviderType = "System.Security.Cryptography.SHA384CryptoServiceProvider, " + AssemblyName_Csp;
+            // string SHA512CryptoServiceProviderType = "System.Security.Cryptography.SHA512CryptoServiceProvider, " + AssemblyName_Csp;
+            // string DpapiDataProtectorType = "System.Security.Cryptography.DpapiDataProtector, " + AssemblyRef.SystemSecurity;
+        }
+
+        private static ConcurrentDictionary<string, Type> CreateTypeNameToTypeMapping()
+        {
+            ConcurrentDictionary<string, Type> typeNameToType = new ConcurrentDictionary<string, Type>(StringComparer.OrdinalIgnoreCase);
+
+            Type HMACMD5Type = typeof(System.Security.Cryptography.HMACMD5);
```
Comment (jkotas) [inline src/System.Security.Cryptography.Algorithms/src/System/Security/Cryptography/CryptoConfig.cs]: It is better to get the type right next to its use. Fetching all types first into locals will result in larger stack frame and bigger native code.

**Comment by bartonjs** [comment]
#39600 has been accepted.  If there're any more gain to be gotten from this PR it would need to be re-evaluated in terms of the new baseline.  Since it'd be a different change with different benchmark justification, I'm closing this PR.

**Pull request closed without merging**